### PR TITLE
[Pablo] Provide LPT when creating pools [CU-3916rnu]

### PR DESCRIPTION
**Pull request by PoisonPhang**
## Issue

[[CU-3916rnu]](https://app.clickup.com/t/3916rnu)

## Additional Changes

Wrapping the LPT ID in an option for future proofing - also has the potential to expose more to testers.

### CODE REVIEW

**Comment by ComposableFiB0t** [comment]
Task linked: [CU-3916rnu Fix ED for LPT [Critical path - blocked add and remove liquidity]](https://app.clickup.com/t/3916rnu)

**Comment by ComposableFiB0t** [comment]
Task linked: [CU-30xvvdg 5.3) 6. Implement Algorithm: Amm::add_liquidity.](https://app.clickup.com/t/30xvvdg)

File: code/parachain/frame/pablo/src/lib.rs [hunk 1]
```diff
@@ -364,7 +364,7 @@ pub mod pallet {
 		#[pallet::weight(T::WeightInfo::create())]
 		pub fn create(origin: OriginFor<T>, pool: PoolInitConfigurationOf<T>) -> DispatchResult {
 			T::PoolCreationOrigin::ensure_origin(origin)?;
-			let _ = Self::do_create_pool(pool)?;
+			let _ = Self::do_create_pool(pool, None)?;
```
Comment (vimukthi-git) [inline code/parachain/frame/pablo/src/lib.rs:367]: Nice :)

File: code/parachain/runtime/picasso/src/lib.rs [hunk 2]
```diff
@@ -252,6 +252,40 @@ impl assets_registry::Config for Runtime {
...
+	pub PabloPalletId: PalletId = PalletId(*b"pal_pblo");
+	pub TWAPInterval: u64 = (MILLISECS_PER_BLOCK as u64) * 10;
+	pub MaxStakingRewardsPools: u32 = 10;
+	pub MaxRewardConfigsPerPool: u32 = 10;
+	pub MaxStakingDurationPresets: u32 = 10;
+	pub PbloAssetId: CurrencyId = CurrencyId::PBLO;
+	pub MsPerBlock: u32 = MILLISECS_PER_BLOCK;
+}
+
+impl pablo::Config for Runtime {
```
Comment (vimukthi-git) [inline code/parachain/runtime/picasso/src/lib.rs:265]: Let's not use PIcasso for this test. We have to remove as any Picasso modification needs to be separately approved as of recent agreements.

**Review by vimukthi-git — COMMENTED** [review]
LGTM overall, but unfortunately we can't use Picasso for testing. Let's use Dali for now and we can introduce Picasso version in a second PR as it's the first time we are adding Pablo to Picasso runtime.

File: code/parachain/runtime/picasso/src/migrations.rs [hunk 3]
```diff
@@ -45,6 +45,67 @@ impl OnRuntimeUpgrade for TechCollectiveRenameMigration {
...
+	use super::*;
+
+	use frame_support::BoundedBTreeMap;
+	use pablo::pallet::PoolInitConfiguration;
+	use sp_runtime::PerThing;
+
+	pub struct PabloPicassoInitialPoolsMigration;
+
+	fn add_initial_pools_to_storage() -> Weight {
+		Pablo::do_create_pool(
```
Comment (vimukthi-git) [inline code/parachain/runtime/picasso/src/migrations.rs:59]: Looks super straight forward when directly using interfaces from Pablo :)

File: code/parachain/runtime/picasso/src/migrations.rs [hunk 4]
```diff
@@ -45,6 +45,67 @@ impl OnRuntimeUpgrade for TechCollectiveRenameMigration {
...
+			create_two_token_pool_config(
+				CurrencyId::PICA,
+				CurrencyId::USDT,
+				Permill::from_percent(50),
+			),
+			Some(CurrencyId::PICA_USDT_LPT),
+		)
+		.expect("Is a valid pool config; QED");
+		// TODO(connor) calculate weight based of read/write count
```
Comment (benluelo) [inline code/parachain/runtime/picasso/src/migrations.rs:78]: it should be possible to write a benchmark for `do_create_pool` when passing an LP token in

File: code/parachain/runtime/picasso/src/migrations.rs [hunk 5]
```diff
@@ -45,6 +45,67 @@ impl OnRuntimeUpgrade for TechCollectiveRenameMigration {
...
+		0
+	}
+
+	fn create_two_token_pool_config(
+		base_asset_id: CurrencyId,
+		quote_asset_id: CurrencyId,
+		base_asset_weigth: Permill,
```
Comment (benluelo) [inline code/parachain/runtime/picasso/src/migrations.rs:85]: weigth -> weight

File: code/parachain/runtime/picasso/src/migrations.rs [hunk 6]
```diff
@@ -45,6 +45,67 @@ impl OnRuntimeUpgrade for TechCollectiveRenameMigration {
...
+		let owner = AccountId::from([0; 32]);
+		let mut assets_weights = BoundedBTreeMap::new();
+		let fee = Permill::from_percent(0);
+
+		assets_weights
+			.try_insert(base_asset_id, base_asset_weigth)
+			.expect("Within BT Bounds; QED");
+		assets_weights
+			.try_insert(quote_asset_id, base_asset_weigth.left_from_one())
+			.expect("Within BT Bounds; QED");
```
Comment (benluelo) [inline code/parachain/runtime/picasso/src/migrations.rs:96]: could be:
```rust
[
  (base_asset_id, base_asset_weigth),
  (quote_asset_id, base_asset_weigth.left_from_one()),
]
.into_iter()
.try_collect()
.expect("Within BT Bounds; QED");
```

File: code/parachain/runtime/picasso/src/migrations.rs (hunk 5 above)
Comment (benluelo) [inline code/parachain/runtime/picasso/src/migrations.rs:85]: should we be using base and quote here?

Comment (vimukthi-git) [reply code/parachain/runtime/picasso/src/migrations.rs:85]: Good catch, we should not use that naming anymore

File: code/parachain/runtime/picasso/src/migrations.rs (hunk 6 above)
Comment (vimukthi-git) [reply code/parachain/runtime/picasso/src/migrations.rs:96]: We can just avoid passing around base/quote and just pass the required assets map I think

Comment (benluelo) [reply code/parachain/runtime/picasso/src/migrations.rs:96]: simple enough to do this as well:
```rust
pub fn dual_asset_pool_weights(
	first_asset: AssetId,
	first_asset_weight: Permill,
	second_asset: AssetId,
) -> BoundedBTreeMap<AssetId, Permill, ConstU32<2>> {
	bounded_btree_map! {
		first_asset => first_asset_weight,
		second_asset => first_asset_weight.left_from_one(),
	}
}
```
Which will panic if the bounds aren't met (stolen from what I'm working on for dex-router  tests)

File: code/parachain/runtime/dali/src/migrations.rs [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+		weights::pablo::WeightInfo::<Runtime>::do_create_pool(pools.len() as u32)
+	}
+
+	fn create_two_token_pool_config(
+		first_asset_id: CurrencyId,
+		second_asset_id: CurrencyId,
+		first_asset_weight: Permill,
+		fee: Permill,
+	) -> PoolInitConfiguration<AccountId, CurrencyId> {
+		let owner = AccountId::from([0; 32]);
```
Comment (PoisonPhang) [inline code/parachain/runtime/dali/src/migrations.rs:74]: Not too sure if the owner should be `0` or someone else

Comment (vimukthi-git) [reply code/parachain/runtime/dali/src/migrations.rs:74]: For dali should not matter

Comment (PoisonPhang) [reply code/parachain/runtime/dali/src/migrations.rs:74]: > We can even set it to the derived account of the Pablo pallet tbh

File: code/parachain/runtime/common/src/fees.rs [hunk 8]
```diff
@@ -73,6 +73,8 @@ impl WellKnownForeignToNativePriceConverter {
 			CurrencyId::kUSD => Some(rational!(15 / 1_000)),
 			CurrencyId::PICA => Some(rational!(1 / 1)),
 			CurrencyId::PBLO => Some(rational!(1 / 1)),
+			CurrencyId::KSM_USDT_LPT => Some(rational!(1 / 1_000_000_000)),
```
Comment (vimukthi-git) [inline code/parachain/runtime/common/src/fees.rs:76]: What would be the ED in PICA based on this value for these LPT tokens?

**Review by vimukthi-git — APPROVED** [review]
LGTM :rocket:

File: code/parachain/runtime/common/src/fees.rs (hunk 8 above)
Comment (vimukthi-git) [reply code/parachain/runtime/common/src/fees.rs:76]: ok so 1 nano PICA

File: code/parachain/frame/pablo/src/weights.rs [hunk 9]
```diff
@@ -21,4 +22,5 @@ impl WeightInfo for () {
 	fn remove_liquidity() -> Weight {10_000}
 	fn buy() -> Weight {10_000}
 	fn swap() -> Weight {10_000}
+	fn do_create_pool() -> Weight {10_000}
```
Comment (dzmitry-lahoda) [inline code/parachain/frame/pablo/src/weights.rs:25]: weird naming prefix `do_`

Comment (benluelo) [reply code/parachain/frame/pablo/src/weights.rs:25]: It's benching `do_create_pool` specifically

File: code/parachain/runtime/dali/src/migrations.rs [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+				lp_token: lp_asset,
+			}
+		}
+	}
+
+	/// Adds pools to Pablo Storage
+	///
+	/// Expects a vec of (pool_init_config, pool_lp_token_id)
+	fn add_initial_pools_to_storage(pools: Vec<PoolCreationInput>) -> Weight {
+		pools.iter().for_each(|pool_creation_input| {
```
Comment (dzmitry-lahoda) [inline code/parachain/runtime/dali/src/migrations.rs:57]: ensure that call to add_initial_pools_to_storage is idempotent

**Review by vimukthi-git — APPROVED** [review]

File: code/parachain/runtime/dali/src/migrations.rs [hunk 11]
```diff
@@ -0,0 +1,208 @@
+use crate::{prelude::*, *};
+
+use pablo_picasso_init_pools::PabloPicassoInitialPoolsMigration;
+
+pub type Migrations = (PabloPicassoInitialPoolsMigration, SchedulerMigrationV3);
+
+// Migration for scheduler pallet to move from a plain Call to a CallOrHash.
```
[the remark this answers is not in the document]
Comment (PoisonPhang) [reply code/parachain/runtime/dali/src/migrations.rs:7]: How so? That seems accurate for the `SchedulerMigrationV3` struct

**Review by KaiserKarel — APPROVED** [review]

File: code/parachain/runtime/dali/src/migrations.rs [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+				lp_token: lp_asset,
+			}
+		}
+	}
+
+	/// Adds pools to Pablo Storage
+	///
+	/// Expects a vec of (pool_init_config, pool_lp_token_id)
+	fn add_initial_pools_to_storage(pools: Vec<PoolCreationInput>) -> Weight {
+		if Pablo::pool_count().is_zero() {
```
Comment (PoisonPhang) [inline code/parachain/runtime/dali/src/migrations.rs:57]: Should be is not zero

**Pull request merged**